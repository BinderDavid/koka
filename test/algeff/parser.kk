effect parse {
  satisfy<b>( pred: string -> maybe<(b,string)> ) : b;
  choice<b>( p1 :  () -> xparse b, p2 : () -> xparse b ) : b
}

alias xparse = <parse,div>

function showP( xs : list<(a,string)>, f : a -> string ) : string {
  xs.showList(fun(r) {
    val (x,rest) = r
    "(" + x.f + "," + rest.show + ")"
  })
}

function showP1( xs : list<(list<int>,string)> ) : string {
  xs.showP( fun(is) { is.showList(show) })
}


val parse = handler(input) {
  return x   -> [(x,input)]
  satisfy(p) -> {
    match(p(input)) {
      Nothing          -> []
      Just((res,rest)) -> resume(rest,res)
    }
  }
  choice(p1,p2) -> {
    val xs = parse(input,p1) + (parse(input,p2) : list<_>)
    xs.concat( fun(xr) {
      val (x,rest) = xr
      resume(rest,x)
    })
  }
}

function fail() {
  satisfy(fun(input) { Nothing })
}

function char(p : char -> bool) : parse char {
  satisfy( fun(input) {
    val sc = input.substr(0,1)
    match (sc.list) {
      [c] | p(c) -> Just((c,input.substr(1)))
      _   -> Nothing 
    }
  })
}

function digit() : parse int {
  int(char(isDigit) - '0')
}

function alpha() : parse int {
  int(char(isAlpha) - 'a')
}

function many(p) {
  manyx([],p)
}

function manyx(acc,p) {
  choice {
    val x = p()
    manyx(Cons(x,acc),p)
  } { acc.reverse } 
}

function manyy(p) {
  choice {
    val x = p()
    val xs = manyy(p)
    Cons(x,xs)
  } { [] } 
}