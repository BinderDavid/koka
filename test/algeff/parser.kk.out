[(7,""),(3,"*3"),(1,"+2*3")]
test/algeff/parser.kk(64, 7): warning: input shadows an earlier local definition or parameter
.Op-fail : forall<a> .op-fail
.Op-flip : .op-flip
.Op-satisfy : forall<a> (pred : (string) -> maybe<(a, string)>) -> .op-satisfy
.Ops-fail : (fail : .op-fail) -> .ops-many
.Ops-flip : (flip : .op-flip) -> .ops-many
.Ops-satisfy: (satisfy : .op-satisfy) -> .ops-parse
.tag-many : string
.tag-op-fail: string
.tag-op-flip: string
.tag-op-satisfy: string
.tag-parse : string
alpha : () -> parse int
char : (p : (char) -> bool) -> parse char
choice : forall<a,e> (p1 : () -> <many|e> a, p2 : () -> <many|e> a) -> <many|e> a
digit : () -> parse int
eager : forall<a,e> (action : () -> <many|e> a) -> e list<a>
eparse : forall<a,e> (p : () -> <many,parse|e> a, input : string) -> e list<(a, string)>
expr : () -> <many,parse,div> int
factor : () -> <many,parse,div> int
fail : forall<a> () -> many a
flip : () -> many bool
main : () -> <div,console> ()
many : forall<a,e> (p : () -> <many,div|e> a) -> <many,div|e> list<a>
many1 : forall<a,e> (p : () -> <many,div|e> a) -> <many,div|e> list<a>
mparse : forall<a,e> (p : () -> <many,parse|e> a, input : string) -> e list<(a, string)>
number : () -> <many,parse,div> int
parse : forall<a,e> (string, () -> <parse,many|e> a) -> <many|e> (a, string)
satisfy : forall<a> (pred : (string) -> maybe<(a, string)>) -> parse a
select : forall<a> (list<a>) -> many a
show-parse : forall<a> (xs : list<(a, string)>, f : (a) -> string) -> string
show-parse-int: (xs : list<(int, string)>) -> string
show-parse-ints: (xs : list<(list<int>, string)>) -> string
solutions : forall<a,e> (action : () -> <many|e> a) -> e list<a>
symbol : (c : char) -> parse ()
term : () -> <many,parse,div> int
test1 : () -> <div,console> ()